### ISSUE TITLE
Labels and IP addresses indexes and selector relationships

### ISSUE THREAD

**Issue by simon3z** [GoogleCloudPlatform/kubernetes#4817]
In a discussion I had yesterday with @bgrant0607 and @lavalamp it came up the need of indexing labels and probably ip addresses. Reference: https://github.com/GoogleCloudPlatform/kubernetes/pull/4482

> We'll still want to do the label indexing, reverse label indexing, lookup by IP address, specification of query URLs in service and controller statuses, field selection for more efficient listing of just metadata, etc.

On top of that I would like also to expose the entity relationship somehow in the api (e.g. /api/v1beta3/relations/{services,replicationcontroller,...})

**Comment by bgrant0607** [comment]
Let's make this issue about indexing.

@simon3z Please discuss the relationship issue in #3676.

**Comment by lavalamp** [comment]
It is on my list of things to do to add a brief sketch here of how I see this working.

**Comment by bgrant0607** [comment]
Possible (mostly complementary) approaches to make the queries more efficient:
* Cache selector results -- replication controllers and endpoint controllers would ensure that selectors expressed through the API were always in cache
* Incrementally update cached selector results upon object creation, update, and deletion (as opposed to cache invalidation)
* Parallelize selector evaluation across objects
* Reduce selector matching to a more efficient representation, such as bit sets and/or bloom filters -- there are a finite number of label keys and values in the system at any given time
* Index labels individually and union/intersect their result sets (again, using an efficient representation, like bit sets)
* Parallelize the intersection approach using a reduction tree (mapreduce!)
* Keep track of the number of matches of each label individual and evaluate most selective requirements first
* Bucket objects by common labels -- this would automatically put all pods managed by a replication controller into the same bucket, for example
* Build an index tree based on increasingly specific subsets of common labels: root would be labels common to all objects, possibly none, and leaves would be the buckets with common labels mentioned above

**Comment by xudifsd** [comment]
@bgrant0607 Is it possible to cache selector result in etcd, and consumers watch its change. incrementally update cached selector is done by apiserver: every time api server add a object with labels it will traverse all selector, and update each one if needed?

**Comment by bgrant0607** [comment]
Points made in #5548: 
* pre-filter by namespace
* watch everything and associate objects with selectors with matching objects in controllers

**Comment by lavalamp** [comment]
Here is a partial design.

For reference, right now our list/watch queries go directly to etcd. Our list goes directly to etcd list, our watch goes directly to watch, with some filtering (we have to watch more--potentially much more--than we strictly would like to).

What we want to make is an object that allows runtime definition of indexes, and supports two basic operations: List and Watch.

## Index definition:

An index is a name and a predicate function. For example, the name might be `"role=frontend"` and the predicate function might be ```func(obj runtime.Object) bool {return labelsOf(obj)["role"] == "frontend"}```.

For every index, we store the key of every object matching the corresponding predicate, that is, a map of index name, to a set of object keys matching the predicate.

I would like to not cache the data for all the objects and store only their key. However, unless someone can think of a solution to the list problem in the next section, it might be necessary. We'll have to do the math and see how reasonable it is to store everything.

It's important that the index have a concept of "how recent am I", and for efficiency, it'd be good if we added/removed items to the indices as the objects change instead of rebuilding the index frequently. We can satisfy both of these requirements by using, as the input to our indexer, a list/watch cycle on etcd. (EtcdIndex AKA ResourceVersion answers the "how recent am I" question.)

We will need to periodically rebuild the index completely (for example, when a new index is added, or if the watch fails/ends and we start a new list/watch cycle). We can do this in the background, while still responding to requests with the last known index

## List

A list request to the indexer might look like, "Give me every pod with labels matching 'role=frontend' and 'spec.host=node15.mycluster.com'." In response to this, the indexer:
1. Determines which of the requested predicates have active indexes.
2. Takes the intersection of all the keys in those indexes,
3. Returns this list of keys (or entire objects, if they're cached), along with the indexer's resourceVersion.

It's then the responsibility of the caller to actually retrieve the objects, and filter them by any unindexed predicates.

Note that, assuming we store keys only and not entire objects, there's an unsolved problem here about returning a list that is complete as of a particular resourceVersion, given that etcd doesn't (yet?) give past history for objects.

Potential optimization: it'd be nice if the indexer tracked requests it'd seen by frequency, so that it could periodically redefine its indexes such that most requests hit an index.

## Watch

We'll define our own N long change history. Every watch request to our apiserver will result in a new watcher being added on to our indexer instead of being plumbed straight to etcd. Adding an item to an index because a predicate matches is very similar to sending an item down a watch because a predicate matches. We can save a bit on computation in the case that multiple indexes/watchers use the same predicate.

When the indexer's own watch expires and it does a re-list, we probably have to kick all downstream watch clients off. However, that would immediately create a thundering heard of list calls. Instead, since we already know exactly what each client wants, we can define a new watch event type, say RESYNC, send the client the new list, and instruct the client to treat that list as if it were the result of a new list command.

**Comment by smarterclayton** [comment]
Makes sense.  We should keep in mind a provision to allow AP consumers to provide the index they read as a precondition for a read and either wait or reject.  This could be at a layer on top of the indexer.

**Comment by derekwaynecarr** [comment]
There is no batch get me these keys request from etcd yet.  I submitted a feature request for that at some point last year.  

https://github.com/coreos/etcd/issues/1282

Does this proposal assume lookup by key is just a cache.Store lookup?  Is this implemented as an enhancement to cache.indexer to ask a caching function to expose a predicate?

**Comment by xudifsd** [comment]
How do you decide which label to index and which not initially? Also, if a selector matches none of indexed label, we have to fall back to list & filter all objs in etcd/cache, right? So I was thinking maybe we could index all the labels objs have.

**Comment by bgrant0607** [comment]
cc @hchoudh

**Comment by smarterclayton** [comment]
One quick question - can you clearly describe what this adds on top of the existing cache.indexer except dynamic on the fly indexing?

**Comment by smarterclayton** [comment]
(aside from watch unification)

**Comment by smarterclayton** [comment]
Also, if we do this for normal read path I think we have to have read precondition

**Comment by lavalamp** [comment]
> Does this proposal assume lookup by key is just a cache.Store lookup? Is this implemented as an enhancement to cache.indexer to ask a caching function to expose a predicate?

Totally up to the implementer whether or not to use cache.Store. It would need some modifications to emit proper resource versions when answering a list.

> How do you decide which label to index and which not initially? Also, if a selector matches none of indexed label, we have to fall back to list & filter all objs in etcd/cache, right? So I was thinking maybe we could index all the labels objs have.

At first we just index entire requests; I think a simple LRU cache is OK for that. Later, when we get really fancy, we can index every label separately and have a super efficient selector lookup function.

> One quick question - can you clearly describe what this adds on top of the existing cache.indexer except dynamic on the fly indexing?

* Source is etcd directly, instead of apiserver
* Multiple object types stored in a single cache (possibly)
* apiserver itself doesn't use any stores (at least not yet)
* Support resourceVersion generation on lists
* Support watch clients
* Support pkg/tools.EtcdHelper interface instead of pkg/client.Interface interface.

> if we do this for normal read path I think we have to have read precondition

To prevent clients from going backwards in time? or something else?

**Comment by jian-he** [comment]
@lavalamp, got a question, is the proposal for adding indexing at the controller side (i.e. the local cache ), or to add indexing at API-server ?

**Comment by anencore94** [comment]
any updates here ?

**Comment by lavalamp** [comment]
I suspect we should merge this issue with #53459. They are different lenses on the same basic deficiency, I think.
